### PR TITLE
rgw: reshard stale instance cleanup

### PR DESCRIPTION
**Pull request by theanalyst**
Depends on https://github.com/ceph/ceph/pull/24406; supersedes https://github.com/ceph/ceph/pull/24592, still need to handle cleanup of metadata keys themselves

Fixes: http://tracker.ceph.com/issues/24082

*edit*- added tracker reference

### CODE REVIEW

**Comment by theanalyst** [comment]
needs to be rebased against current version of https://github.com/ceph/ceph/pull/24406

File: src/rgw/rgw_bucket.cc [hunk 1]
```diff
@@ -1659,6 +1660,219 @@ int RGWBucketAdminOp::set_quota(RGWRados *store, RGWBucketAdminOpState& op_state
   return bucket.set_quota(op_state);
 }
 
+inline std::string bucket_instance_name(const std::string&bucket, const std::string& bucket_id)
```
[the remark this answers is not in the document]
Comment (theanalyst) [reply src/rgw/rgw_bucket.cc]: good catch

File: src/rgw/rgw_bucket.cc [hunk 2]
```diff
@@ -1659,6 +1660,208 @@ int RGWBucketAdminOp::set_quota(RGWRados *store, RGWBucketAdminOpState& op_state
...
+  // Now we have a bucket with instances where the reshard status is none, this
+  // usually happens when the reshard process couldn't complete, lockdown the
+  // bucket and walk through these instances to make sure no one else interferes
+  // with these
+  {
+    RGWBucketReshardLock reshard_lock(store, cur_bucket_info, true);
+    r = reshard_lock.lock();
+    if (r < 0) {
+      // most likely bucket is under reshard, return the sureshot stale instances
+      ldout(store->ctx(), 10) << __func__
```
Comment (ivancich) [inline src/rgw/rgw_bucket.cc]: Needs a space and change of "likey" to "likely".

File: src/rgw/rgw_bucket.cc [hunk 3]
```diff
@@ -1659,6 +1660,208 @@ int RGWBucketAdminOp::set_quota(RGWRados *store, RGWBucketAdminOpState& op_state
...
+{
+  std::string marker;
+  void *handle;
+  Formatter *formatter = flusher.get_formatter();
+  static constexpr auto default_max_keys = 1000;
+
+  int ret = store->meta_mgr->list_keys_init("bucket.instance", marker, &handle);
+  if (ret < 0) {
+    cerr << "ERROR: can't get key: " << cpp_strerror(-ret) << std::endl;
+    return -ret;
```
Comment (ivancich) [inline src/rgw/rgw_bucket.cc]: Negative or positive return value?

File: src/rgw/rgw_bucket.cc [hunk 4]
```diff
@@ -1659,6 +1660,208 @@ int RGWBucketAdminOp::set_quota(RGWRados *store, RGWBucketAdminOpState& op_state
...
+
+  formatter->open_array_section("keys");
+
+  do {
+    list<std::string> keys;
+
+    ret = store->meta_mgr->list_keys_next(handle, default_max_keys, keys, &truncated);
+    if (ret < 0 && ret != -ENOENT) {
+      cerr << "ERROR: lists_keys_next(): " << cpp_strerror(-ret) << std::endl;
+      return -ret;
```
Comment (ivancich) [inline src/rgw/rgw_bucket.cc]: See above.

File: src/rgw/rgw_bucket.cc [hunk 5]
```diff
@@ -1659,6 +1660,208 @@ int RGWBucketAdminOp::set_quota(RGWRados *store, RGWBucketAdminOpState& op_state
...
+    } if (ret != -ENOENT) {
+      // partition the list of buckets by buckets as the listing is un sorted,
+      // since it would minimize the reads to bucket_info
+      std::unordered_map<std::string, std::vector<std::string>> bucket_instance_map;
+      for (auto &key: keys) {
+        if(auto pos = key.find(':'); pos != std::string::npos)
```
Comment (ivancich) [inline src/rgw/rgw_bucket.cc]: Another if-with-initializer.

File: src/rgw/rgw_bucket.cc [hunk 6]
```diff
@@ -1659,6 +1660,208 @@ int RGWBucketAdminOp::set_quota(RGWRados *store, RGWBucketAdminOpState& op_state
...
+{
+  int max_shards = (bucket_info.num_shards > 0 ? bucket_info.num_shards : 1);
+  for (int i = 0; i < max_shards; i++) {
+    RGWRados::BucketShard bs(store);
+    int shard_id = (bucket_info.num_shards > 0  ? i : -1);
+    int ret = bs.init(bucket_info.bucket, shard_id, nullptr);
+    if (ret < 0) {
+      cerr << "ERROR: bs.init(bucket=" << bucket_info.bucket << ", shard=" << shard_id
+           << "): " << cpp_strerror(-ret) << std::endl;
+      return -ret;
```
[the remark this answers is not in the document]
Comment (ivancich) [reply src/rgw/rgw_bucket.cc]: I haven't traced these all the way out to see how they might end up in json. I do know other rgw code mixed positive error return values with negative error return values. @cbodley is there a standard?

File: src/rgw/rgw_bucket.cc [hunk 7]
```diff
@@ -1659,6 +1660,211 @@ int RGWBucketAdminOp::set_quota(RGWRados *store, RGWBucketAdminOpState& op_state
...
+                                            RGWFormatterFlusher& flusher)
+{
+  auto process_f = [](const bucket_instance_ls& lst,
+                      Formatter *formatter,
+                      RGWRados *store){
+                     for (const auto &binfo: lst) {
+                       int ret = purge_bucket_instance(store, binfo);
+                       if (ret == 0){
+                         auto md_key = "bucket.instance:" + binfo.bucket.get_key();
+                         ret = store->meta_mgr->remove(md_key);
```
Comment (ivancich) [inline src/rgw/rgw_bucket.cc]: RGWMetadataManager::remove returns negative error codes....

File: src/rgw/rgw_bucket.cc [hunk 8]
```diff
@@ -1659,6 +1660,211 @@ int RGWBucketAdminOp::set_quota(RGWRados *store, RGWBucketAdminOpState& op_state
...
+                       }
+                       formatter->open_object_section("delete_status");
+                       formatter->dump_string("bucket_instance", binfo.bucket.get_key());
+                       formatter->dump_int("status", ret);
```
Comment (ivancich) [inline src/rgw/rgw_bucket.cc]: ... so if you want this to be a positive value, this might be a better place to negate (from negative to positive)?

Comment (theanalyst) [reply src/rgw/rgw_bucket.cc]: ack

**Comment by ivancich** [comment]
Other than some needed squashing of commits, it looks good!

**Comment by theanalyst** [comment]
jenkins test this please

File: src/rgw/rgw_bucket.cc [hunk 9]
```diff
@@ -1659,6 +1660,210 @@ int RGWBucketAdminOp::set_quota(RGWRados *store, RGWBucketAdminOpState& op_state
...
+  if (other_instances.empty()) {
+    return;
+  }
+
+  // Now we have a bucket with instances where the reshard status is none, this
+  // usually happens when the reshard process couldn't complete, lockdown the
+  // bucket and walk through these instances to make sure no one else interferes
+  // with these
+  {
+    RGWBucketReshardLock reshard_lock(store, cur_bucket_info, true);
```
Comment (ivancich) [inline src/rgw/rgw_bucket.cc]: It looks like this code assumes that when reshard_lock goes out of scope that the lock will be unlocked automatically (via destructor). That's not currently the case, and you'll want to unlock at the bottom of the scope.

Comment (ivancich) [reply src/rgw/rgw_bucket.cc]: So what that means is that the lock will remain until it expires, which might be 6 minutes by default.

File: src/rgw/rgw_bucket.cc [hunk 10]
```diff
@@ -1659,6 +1660,210 @@ int RGWBucketAdminOp::set_quota(RGWRados *store, RGWBucketAdminOpState& op_state
...
+    r = reshard_lock.lock();
+    if (r < 0) {
+      // most likely bucket is under reshard, return the sureshot stale instances
```
Comment (ivancich) [inline src/rgw/rgw_bucket.cc]: I'm unfamiliar with what "sureshot" means and not in this context. And wikipedia didn't clarify (see https://en.wikipedia.org/wiki/Sureshot). Perhaps clarify or consider alternatives.

File: src/rgw/rgw_admin.cc [hunk 11]
```diff
@@ -977,8 +982,13 @@ static int get_cmd(const char *cmd, const char *prev_cmd, const char *prev_prev_
       return OPT_MFA_CHECK;
     if (strcmp(cmd, "resync") == 0)
       return OPT_MFA_RESYNC;
+  } else if ((prev_prev_cmd && strcmp(prev_prev_cmd, "reshard") == 0) &&
+	     (strcmp(prev_cmd, "stale-instances") == 0)) {
+    if ((strcmp(cmd, "list")) == 0)
```
Comment (ivancich) [inline src/rgw/rgw_admin.cc]: Extra set of parentheses.

File: src/rgw/rgw_bucket.cc (hunk 9 above)
Comment (ivancich) [reply src/rgw/rgw_bucket.cc]: Sorry I didn't notice this on my first review....

File: src/rgw/rgw_bucket.cc (hunk 10 above)
Comment (theanalyst) [reply src/rgw/rgw_bucket.cc]: I meant more like https://en.oxforddictionaries.com/definition/sure-shot (but I see now it is more in an informal sense, I'll reword)

File: src/rgw/rgw_bucket.cc (hunk 9 above)
Comment (theanalyst) [reply src/rgw/rgw_bucket.cc]: Shouldn't the scope_guard take care of this a few lines below where the unlock will be called by the destructor.  Though I'm okay with an explicit unlock below as well

File: src/rgw/rgw_admin.cc (hunk 11 above)
Comment (theanalyst) [reply src/rgw/rgw_admin.cc]: good catch

File: src/rgw/rgw_bucket.cc (hunk 10 above)
Comment (ivancich) [reply src/rgw/rgw_bucket.cc]: Cool; I didn't know that.

File: src/rgw/rgw_bucket.cc (hunk 9 above)
Comment (cbodley) [reply src/rgw/rgw_bucket.cc]: i like the scope_guard just for exception safety, since the insert() could potentially throw bad_alloc

Comment (ivancich) [reply src/rgw/rgw_bucket.cc]: You are correct; my mistake. I saw that last week but then forgot about it as I was doing the backport.

**Comment by ivancich** [comment]
So I first ran a downstream version that did not have my auto-cleanup
code to create some test cases.

1. I created a bucket.
2. I did a successful reshard.
3. I did a reshard that took too long and didn't finish.
4. I did a reshard that crashed part-way through.

So I should have 4 bucket infos -- the initial one and the three from the three reshards (1 successful, 2 unsuccessful).

When I do a `bin/rados -p default.rgw.meta -N root ls` I get:

    .bucket.meta.reshard-test:2ede1521-42f2-4d49-b11d-bdb1364a2223.14101.1
    reshard-test
    .bucket.meta.reshard-test:2ede1521-42f2-4d49-b11d-bdb1364a2223.14097.1
    .bucket.meta.reshard-test:2ede1521-42f2-4d49-b11d-bdb1364a2223.24099.1
    .bucket.meta.reshard-test:2ede1521-42f2-4d49-b11d-bdb1364a2223.34099.1

When I do a `bin/rados -p default.rgw.buckets.index ls` I get:

    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.24099.1.15
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.34099.1.12
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.14101.1.0
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.34099.1.3
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.34099.1.13
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.14101.1.7
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.24099.1.6
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.34099.1.19
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.14101.1.4
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.24099.1.8
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.34099.1.1
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.24099.1.12
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.14101.1.6
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.14101.1.3
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.24099.1.11
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.14097.1
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.24099.1.9
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.34099.1.0
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.34099.1.11
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.34099.1.10
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.34099.1.6
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.14101.1.1
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.14101.1.2
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.34099.1.14
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.34099.1.2
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.24099.1.4
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.24099.1.7
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.34099.1.9
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.24099.1.3
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.14101.1.5
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.34099.1.8
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.24099.1.14
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.34099.1.7
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.34099.1.5
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.34099.1.15
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.34099.1.4
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.24099.1.5
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.24099.1.1
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.24099.1.2
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.34099.1.18
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.24099.1.10
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.14101.1.8
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.24099.1.0
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.34099.1.17
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.34099.1.16
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.24099.1.13

When I then do a `bin/radosgw-admin reshard stale-instances list` I get:

    [
        "reshard-test:2ede1521-42f2-4d49-b11d-bdb1364a2223.14097.1"
    ]

So it only picked up one of the three possible bucket info objects (the original bucket info that was left behind after the successful reshard), so I'm guessing it will only clean up the one. Now I'll try to clean up.

When I then do a `bin/radosgw-admin reshard stale-instances rm` I get:

    [
        {
            "bucket_instance": "reshard-test:2ede1521-42f2-4d49-b11d-bdb1364a2223.14097.1",
            "status": 0
        }
    ]

Now when I do a `bin/rados -p default.rgw.meta -N root ls` I get:

    .bucket.meta.reshard-test:2ede1521-42f2-4d49-b11d-bdb1364a2223.14101.1
    reshard-test
    .bucket.meta.reshard-test:2ede1521-42f2-4d49-b11d-bdb1364a2223.24099.1
    .bucket.meta.reshard-test:2ede1521-42f2-4d49-b11d-bdb1364a2223.34099.1

So in fact, the other two were left behind.

And now when I do a `bin/rados -p default.rgw.buckets.index ls` I get:

    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.24099.1.15
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.34099.1.12
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.14101.1.0
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.34099.1.3
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.34099.1.13
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.14101.1.7
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.24099.1.6
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.34099.1.19
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.14101.1.4
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.24099.1.8
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.34099.1.1
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.24099.1.12
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.14101.1.6
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.14101.1.3
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.24099.1.11
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.24099.1.9
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.34099.1.0
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.34099.1.11
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.34099.1.10
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.34099.1.6
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.14101.1.1
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.14101.1.2
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.34099.1.14
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.34099.1.2
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.24099.1.4
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.24099.1.7
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.34099.1.9
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.24099.1.3
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.14101.1.5
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.34099.1.8
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.24099.1.14
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.34099.1.7
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.34099.1.5
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.34099.1.15
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.34099.1.4
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.24099.1.5
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.24099.1.1
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.24099.1.2
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.34099.1.18
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.24099.1.10
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.14101.1.8
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.24099.1.0
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.34099.1.17
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.34099.1.16
    .dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.24099.1.13
                    
Note: ".dir.2ede1521-42f2-4d49-b11d-bdb1364a2223.14097.1" is missing. But as expected, it didn't clean up the shards from the failed attempts at resharding.

So I don't think the listing and delete are finding all the bucket infos and shards that it could.

**Comment by theanalyst** [comment]
When you attempted the cleanup was the bucket still in status 1, since the reshard lock is held for sometime usually the bucket cleanup doesn't pick up the stale instances yet until we can lock. I used to check the path we're exiting with some debug statements in the reshard stale-cleanup path. If the bucket is currently not resharding anymore, does the stale-instances cleanup still not pick these up?

**Comment by ivancich** [comment]
@theanalyst Before these tests I set the lock duration to 30 seconds, so I could also insert a 30-second sleep to cause the first failure (unable to renew lock in time). The cluster is still running, and I just did a `bin/radosgw-admin reshard stale-instances list` and the output is empty (except for a pair of square brackets).

When I just did a `bin/radosgw-admin metadata get bucket.instance:reshard-test:2ede1521-42f2-4d49-b11d-bdb1364a2223.14101.1` I got:

```
{
    "key": "bucket.instance:reshard-test:2ede1521-42f2-4d49-b11d-bdb1364a2223.14101.1",
    "ver": {
        "tag": "uBmRxLki5fMDFXAr93SxuEN8",
        "ver": 5
    },
    "mtime": "2018-11-26 21:04:39.311352Z",
    "data": {
        "bucket_info": {
            "bucket": {
                "name": "reshard-test",
                "marker": "2ede1521-42f2-4d49-b11d-bdb1364a2223.14097.1",
                "bucket_id": "2ede1521-42f2-4d49-b11d-bdb1364a2223.14101.1",
                "tenant": "",
                "explicit_placement": {
                    "data_pool": "",
                    "data_extra_pool": "",
                    "index_pool": ""
                }
            },
            "creation_time": "2018-11-26 20:45:56.740606Z",
            "owner": "testid",
            "flags": 0,
            "zonegroup": "23df11e2-d71b-49f5-affd-7f8c952a7fa2",
            "placement_rule": "default-placement",
            "has_instance_obj": "true",
            "quota": {
                "enabled": false,
                "check_on_raw": false,
                "max_size": -1,
                "max_size_kb": 0,
                "max_objects": -1
            },
            "num_shards": 9,
            "bi_shard_hash_type": 0,
            "requester_pays": "false",
            "has_website": "false",
            "swift_versioning": "false",
            "swift_ver_location": "",
            "index_type": 0,
            "mdsearch_config": [],
            "reshard_status": 1,
            "new_bucket_instance_id": "2ede1521-42f2-4d49-b11d-bdb1364a2223.34099.1"
        },
        "attrs": [
            {
                "key": "user.rgw.acl",
                "val": "AgKNAAAAAwIXAAAABgAAAHRlc3RpZAkAAABNLiBUZXN0ZXIEA2oAAAABAQAAAAYAAAB0ZXN0aWQPAAAAAQAAAAYAAAB0ZXN0aWQFAzsAAAACAgQAAAAAAAAABgAAAHRlc3RpZAAAAAAAAAAAAgIEAAAADwAAAAkAAABNLiBUZXN0ZXIAAAAAAAAAAAAAAAAAAAAA"
            },
            {
                "key": "user.rgw.idtag",
                "val": ""
            }
        ]
    }
}
```

**Comment by ivancich** [comment]
@mattbenjamin @cbodley I worked with @theanalyst to continue my tests. Apparently the issue was that my induced crash left the reshard status as 1 instead of 0, which prevented this code from picking up the incomplete bucket infos. Once I executed another successful reshard, the status went to 0, and the list and delete (clean-up) commands in this PR worked.

**Comment by ivancich** [comment]
I'll do another test like my previous but leave out the induced crash to see what happens then.

**Comment by theanalyst** [comment]
Right now in dynamic resharding while processing logshards (`process_single_logshard`), we check that the `new_instance_id` is empty, for crashes and exits, we'd leave the bucket with reshard status of 1.

What should be the strategy on these buckets, is the admin expected to check for reshard list issue either a new `reshard add` with a changed count or process this with offline resharding with `bucket reshard`

**Comment by ivancich** [comment]
So I did another test where I omitted the induced crash and everything worked well. It would be nice if the system better recovered from a crash, but those are rare events. Perhaps we can address that issue in a future PR. But I think this PR works as expected and is a great addition to radosgw-admin.

**Comment by theanalyst** [comment]
Thanks, I'll schedule a test in teuthology

**Comment by smithfarm** [comment]
jenkins test docs

**Comment by theanalyst** [comment]
tests: http://pulpito.ceph.com/abhi-2018-11-27_17:52:33-rgw-wip-abhi-testing-2018-11-27-1654-distro-basic-smithi/

multisite failures: @cbodley  mentioned these will be fixed by https://github.com/ceph/ceph/pull/25123

valgrind failures on RGWSI_shutdown: fixed by https://github.com/ceph/ceph/pull/25002 which wasnt in integration branch. Should we retest these again?

**Comment by cbodley** [comment]
> Should we retest these again?

nope, this doesn't touch anything outside of the new radosgw-admin commands, so there's little risk of regression

**Comment by smithfarm** [comment]
Is this a candidate for backport? The backport field in http://tracker.ceph.com/issues/24082 appears to be empty.

**Comment by ivancich** [comment]
@smithfarm It is and I've updated the tracker. Thanks!

**Pull request merged**